pilot/pkg/serviceregistry/kube/cache.go

```diff
@@ -45,18 +46,44 @@ func newPodCache(ch cacheHandler) *PodCache {
 		pod := *obj.(*v1.Pod)
 		ip := pod.Status.PodIP
 		if len(ip) > 0 {
+			key := KeyFunc(pod.Name, pod.Namespace)
 			switch ev {
-			case model.EventAdd, model.EventUpdate:
-				out.keys[ip] = KeyFunc(pod.Name, pod.Namespace)
+			case model.EventAdd:
+				switch pod.Status.Phase {
+				case v1.PodPending, v1.PodRunning:
+					// add to cache if the pod is running or pending
+					out.keys[ip] = key
+				}
+			case model.EventUpdate:
+				switch pod.Status.Phase {
+				case v1.PodPending, v1.PodRunning:
+					// add to cache if the pod is running or pending
+					out.keys[ip] = key
+				default:
```

Comment:
why would we delete in PodSucceeded case ?
even Unknown seems dangerous to delete as it could be transient

Reply:
The invariant is that the pod is in the cache only if it is pending or running. We can argue whether the invariant is not right, but code just does what it is supposed to.

Reply:
Ok, bummer, I thought I'd win the prize for noticing missing phases, but... what is this cache used for then ?
(why is it useful to cache only the pods in those 2 states)

Reply:
ah - I didn't read carefully - I assumed "Succeeded" meant "Fully Running" (all containers) but it actually seem to mean Ended (who picked that name...) 

> Succeeded: All Containers in the Pod have *terminated* in success, and will not be restarted.

what is terminated in success vs terminated unsuccessfully ?

Reply:
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/ if anyone else is curious

back to Unknown - shouldn't you keep Unknowns until they come back ? that should avoid 503s no ?

Reply:
Unknown pod seems to indicate some underlying failure in k8s, right? 503s in the issue are happening in normal operation, since some terminated pod evicted a running pod from the cache. The cache is used to associated pods with services by IP. The proxy config for that IP uses this.

Reply:
yeah as I wrote in the main comment line,   k8s server may show unknown transiently for something that still works fine envoy-envoy (maybe even already established connections) - if the removal of that ip from the cache means envoy will drop that endpoint I think we're creating potential errors when we don't need to ? (we should let the normal dead svc detection figure out if point to point it is still working or not or if that was just a glitch in k8s side)